### PR TITLE
Replace moment with date-fns

### LINKED ISSUE #4095: Dashboard: Created/last modified relative strings are sometimes broken
## Bug Description

The first story was modified a few minutes ago.

## Expected Behaviour

It should print the correct relative time.

## Steps to Reproduce

The screenshot is from staging (I'm in UTC+2).

## Screenshots

image

## Additional Context

- Plugin Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by swissspidy**
## Summary

This is a follow-up to #3186, #3244, #3247, and #4082.

This PR solves a few issues...

**Timezone issues in date formatting**

Addresses #4095 by revamping how we deal with timezones from start to finish.

**Usage of `@wordpress/date`**

We currently use the `@wordpress/date` package for date formatting in the editor.

This causes a couple of issues:

* We load it as an **external** (meaning it is not part of the actual bundle) because it is GPL licensed...
* ...which means it itself loads `moment` as an external since it's a dependency.
* An additional hurdle for decoupling the app from WordPress (#2918)

**Moment.js is (too) large**

We only need a small set of features for date-related functionality in the editor/dashboard. Moment.js is too big for that, and there are more lightweight alternatives available.

This PR reduces `assets/js/stories-dashboard.js` bundle size **by 20KB**.

While `assets/js/edit-story.js` is getting 11 KB bigger, we now prevent WordPress from loading `@wordpress/date` (200KB) as an external script, which in turn loads moment.js as an external script (60KB). So in total, we now load **250 KB less JS** for the editor!

**Moment.js is in maintenance mode**

As per https://momentjs.com/docs/#/-project-status/, Moment is considered a legacy project in maintenance mode. It is not dead, but it is indeed _done_.

Time to use more modern and up-to-date alternatives!

**Lack of translation**

Phrases like `%s ago` and `%s from now` that can be seen in the dashboard (using `getRelativeDisplayDate`) are currently not translatable.

This PR fixes this in a clever way by using all the locale information provided by the platform (WordPress) to create a new locale for date-fns with all the necessary translations.

This way we avoid importing a ton of separate translation files.

## Relevant Technical Choices

* uses `date-fns` and `date-fns-tz`, moving all date logic to `assets/src/date`

## To-do

* [x] Tests

## User-facing changes

Some minor wording changes when displaying relative dates (e.g.  "1 hour ago"), but otherwise no user-facing changes.

## Testing Instructions

- Editing media in the editor should still show the creation date.
- Dates in the dashboard should still be shown correctly.
- `wp-includes/js/dist/date.min.js` is no longer loaded in the editor

---

Fixes #4095

### CODE REVIEW

File: assets/src/dashboard/app/api/useStoryApi.js [hunk 1]
```diff
@@ -220,8 +220,8 @@ const useStoryApi = (dataAdapter, { editStoryURL, storyApi }) => {
             title: title,
             author: author || 1,
             slug: title,
-            date: created ? created.format() : Date.now().toString(),
-            modified: modified ? modified.format() : Date.now().toString(),
+            date: created ? created : Date.now().toString(),
+            modified: modified ? modified : Date.now().toString(),
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/api/useStoryApi.js]: could sub out the ternary pattern for just an or `modified || Date.now().toString()`

**Review by BrittanyIRL — COMMENTED** [review]
I'm very on board with updating to date-fns (especially now that [moment's in maintenance](https://momentjs.com/docs/#/-project-status/))! I like the update to use the config within /Date  to reference the locale and just pass the dates themselves through. I went through and tested the dashboard - swapping timezones and editing and publishing to check on continuity of relative time displayed and everything looks good. Tests and storybook work as expected.

**Comment by swissspidy** [comment]
@BrittanyIRL @barklund @spacedmonkey Since 1.0.0 is now out of the door, I would love to get this in & get your review before that.

I like to look at it as _performance optimization & refactoring that happens to fix some bugs_

File: package.json [hunk 2]
```diff
@@ -21,30 +21,27 @@
...
   },
   "dependencies": {
-    "@wordpress/api-fetch": "^3.19.1",
     "@wordpress/block-editor": "^4.5.0",
     "@wordpress/blocks": "^6.22.0",
     "@wordpress/components": "^10.2.0",
     "@wordpress/compose": "^3.20.1",
     "@wordpress/data": "^4.23.1",
-    "@wordpress/date": "^3.11.0",
+    "@wordpress/api-fetch": "^3.19.1",
```
Comment (barklund) [inline package.json]: Why is this one moved to this place? Seems un-alphabetic (anarcho-alphabetic?) 😄 

Also, I have `@wordpress/date` removed in my local `package-lock.json` but not in the committed one - should I just ignore that or?

File: assets/src/date/test/toUTCDate.js [hunk 3]
```diff
@@ -17,19 +17,26 @@
 /**
  * External dependencies
  */
-import moment from 'moment-timezone';
+import MockDate from 'mockdate';
```
Comment (barklund) [inline assets/src/date/test/toUTCDate.js:20]: Wait, why does git think this is a moved file? They have nothing in common except that mandatory license which goes in every file!

File: assets/src/edit-story/components/inspector/document/publish/publishTime.js [hunk 4]
```diff
@@ -89,9 +89,14 @@ function PublishTime() {
       updateStory,
     })
   );
-  const { timeFormat } = useConfig();
+  const {
+    locale: { timeFormat },
+  } = useConfig();
   const use12HourFormat = is12Hour(timeFormat);
```
Comment (barklund) [inline assets/src/edit-story/components/inspector/document/publish/publishTime.js]: This function (`is12hour`) is only ever used here - if we just make it stateful as well (depending on the settings inside `date` package), we can remove this dependency on config here.

And then we can completely remove `locale` from `config` altogether, as this appears to be the only instance where anything from the `locale` property is used through `useConfig()`.

File: assets/src/edit-story/components/inspector/document/test/publish.js [hunk 5]
```diff
@@ -47,7 +47,7 @@ function setupPanel(
     actions: { updateStory },
   };
 
-  const config = { capabilities };
+  const config = { capabilities, locale: { timeFormat: 'g:i a' } };
```
Comment (barklund) [inline assets/src/edit-story/components/inspector/document/test/publish.js]: Why is this needed? It seems like it's not used anywhere - it only appears to have been in use in [`mediaEditDialog#L148`](https://github.com/google/web-stories-wp/pull/4484/files#diff-766a8c842828bdfef73a615b3c39d722L148), which has been removed. Similarly in the next test file.

Ah, found the other instance in `publishTime`, but see [my comment there](https://github.com/google/web-stories-wp/pull/4484/files#diff-18b1cbba739ad149d9329be20614ffc0R95).

File: assets/src/date/index.js [hunk 6]
```diff
@@ -23,15 +23,13 @@
...
-export { isToday } from './isToday';
-export { isYesterday } from './isYesterday';
-export { getDateObjectWithTimezone } from './getDateObjectWithTimezone';
-export { getTimeFromNow } from './getTimeFromNow';
-export { getRelativeDisplayDate } from './getRelativeDisplayDate';
+export { getSettings, updateSettings } from './settings';
+export { default as format } from './format';
+export { default as getRelativeDisplayDate } from './getRelativeDisplayDate';
+export { default as formatDate } from './formatDate';
+export { default as formatTime } from './formatTime';
```
Comment (barklund) [inline assets/src/date/index.js:33]: I don't know how I feel about all these functions being _impure_ (stateful), but I guess it's the easiest way to do this without having to send the same formatting parameters everywhere.

File: assets/src/dashboard/components/cardGridItem/stories/index.js [hunk 7]
```diff
@@ -67,7 +66,7 @@ export const _default = () => {
         <CardTitle
           title="How to be a leader in the apocalpyse"
           author="Rick Grimes"
-          displayDate={moment('04-04-2020').format('MM-DD-YYYY')}
+          displayDate="04-19-2020"
```
Comment (barklund) [inline assets/src/dashboard/components/cardGridItem/stories/index.js]: I know the component just shows the date that's being passed to it as a string, but it still grinds me a bit that it's in the weird (US) order here.

**Review by barklund — APPROVED** [review]
I have reviewed the (rather limited) usage of date formatting in the `edit-story` package and it all seems to hold up.

I have a few general comments and suggestions but otherwise, this looks good.

There is quite a bit more date formatting in the `dashboard` package, so we'll need a thorough review from @BrittanyIRL too!

File: assets/src/dashboard/karma/fixture.js [hunk 8]
```diff
@@ -41,10 +41,13 @@ const defaultConfig = {
...
-  timeFormat: 'g:i a',
-  gmtOffset: -4,
-  timezone: 'America/New_York',
+  locale: {
+    dateFormat: 'F j, Y',
+    timeFormat: 'g:i a',
+    gmtOffset: -4,
+    timezone: 'America/New_York',
+    weekStartsOn: 0,
+  },
```
Comment (BrittanyIRL) [inline assets/src/dashboard/karma/fixture.js:50]: 👍

File: assets/src/date/test/formatTime.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  beforeEach(() => {
+    MockDate.set('2020-07-15T22:47:26+00:00');
+  });
+
+  afterEach(() => {
+    MockDate.reset();
+    resetSettings();
+  });
+
+  it('should return 10:47 pm with no formatting options', () => {
```
Comment (BrittanyIRL) [inline assets/src/date/test/formatTime.js]:
```suggestion
  it('should return 10:47 am with no formatting options', () => {
```

File: assets/src/date/test/formatTime.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      timeFormat: 'g:i A',
+      timezone: 'America/Los_Angeles',
+    });
+    const dateString = '2020-05-02T17:47:26';
+    const formattedTime = formatTime(dateString);
+
+    expect(formattedTime).toStrictEqual('5:47 PM');
+  });
+
+  it('should return 10:47 with H:i formatting options', () => {
```
Comment (BrittanyIRL) [inline assets/src/date/test/formatTime.js]:
```suggestion
  it('should return 20:23 with H:i formatting options', () => {
```

File: includes/Locale.php [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+			'dateFormat'       => get_option( 'date_format' ),
+			'timeFormat'       => get_option( 'time_format' ),
+			'gmtOffset'        => get_option( 'gmt_offset' ),
+			'timezone'         => get_option( 'timezone_string' ),
+			'months'           => array_values( $wp_locale->month ),
+			'monthsShort'      => array_values( $wp_locale->month_abbrev ),
+			'weekdays'         => array_values( $wp_locale->weekday ),
+			'weekdaysShort'    => array_values( $wp_locale->weekday_abbrev ),
+			'weekdaysInitials' => array_values( $wp_locale->weekday_initial ),
+			'weekStartsOn'     => (int) get_option( 'start_of_week', 0 ),
```
Comment (BrittanyIRL) [inline includes/Locale.php:56]: 👏  I like this organization to not clutter up the config with each of these individually. Having them nested is really great.

File: package.json [hunk 12]
```diff
@@ -21,30 +21,27 @@
...
     "draft-js-import-html": "^1.4.1",
     "draftjs-filters": "^2.3.0",
     "flagged": "^2.0.1",
     "history": "^5.0.0",
-    "mockdate": "^3.0.2",
-    "moment": "^2.27.0",
-    "moment-timezone": "^0.5.31",
     "mousetrap": "^1.6.5",
     "patch-package": "^6.2.2",
-    "php-date-formatter": "^1.3.6",
```
Comment (BrittanyIRL) [inline package.json:47]: 🎆 💯 💥  Stoked these can all go away!

**Review by BrittanyIRL — APPROVED** [review]
This looks great and works well for me. I spent more time updating and creating stories and modifying in the dashboard as well from a variety of timezones and formats. Tests run locally for me. Storybook is good. 

Your PR summary here is really great - it's really clear that a lot of detail went into this, thank you for being so thorough. 

LGTM! 🚀 

I don't think this has anything to do with replacing moment with date-fns but I wanted to mention it for posterity - happy to make a bug ticket, lemme know. 

The publishing date/time dropdown in the editor doesn't respect the date format settings.
Screen Shot 2020-09-24 at 12 59 57 PM

Screen Shot 2020-09-24 at 1 00 17 PM

File: package.json (hunk 2 above)
Comment (swissspidy) [reply package.json]: Hmm definitely odd, will double check again.

File: assets/src/date/index.js (hunk 6 above)
Comment (swissspidy) [reply assets/src/date/index.js:33]: It's certainly not the best approach, but I think once this is merged things are in a better shape should there be a need for further refactoring.

**Comment by swissspidy** [comment]
> The publishing date/time dropdown in the editor doesn't respect the date format settings.

The dropdown uses `__( 'd/m/Y ' )`, probably to make sure the format is short enough to fit into the dropdown. The full month name would be too long, for example.

**Pull request merged**